pkg/controller/rolloutblock/rolloutblock_controller.go

```diff
@@ -0,0 +1,313 @@
+
+	// maxRetries is the number of times an RolloutBlock will be retried before
+	// we drop it out of the rolloutblock workqueue. The number is chosen with
+	// the default rate limiter in mind. This results in the following backoff
+	// times: 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s,
+	// 5.1s, 10.2s.
+	maxRetries = 11
+)
+
+// Controller is a Kubernetes controller that updates rolloutblock objects.
+// The RolloutBlock objects' main objective is to block application and releases rollout during
+// an outage.
+//
+// RolloutBlock Controller has one primary workqueues: a rolloutblock updater
+// object queue.
+type Controller struct {
+	shipperClientset clientset.Interface
+	recorder         record.EventRecorder
+
+	applicationLister  shipperlisters.ApplicationLister
+	applicationsSynced cache.InformerSynced
+
+	releaseLister  shipperlisters.ReleaseLister
+	releasesSynced cache.InformerSynced
+
+	rolloutBlockLister shipperlisters.RolloutBlockLister
+	rolloutBlockSynced cache.InformerSynced
+
+	rolloutblockWorkqueue workqueue.RateLimitingInterface
+}
+
+// NewController returns a new RolloutBlock controller.
+func NewController(
+	shipperClientset clientset.Interface,
+	informerFactory shipperinformers.SharedInformerFactory,
+	recorder record.EventRecorder,
+) *Controller {
+	applicationInformer := informerFactory.Shipper().V1alpha1().Applications()
+	releaseInformer := informerFactory.Shipper().V1alpha1().Releases()
+	rolloutBlockInformer := informerFactory.Shipper().V1alpha1().RolloutBlocks()
+
+	glog.Info("Building a RolloutBlock controller")
+
+	controller := &Controller{
+		recorder:         recorder,
+		shipperClientset: shipperClientset,
+
+		applicationLister:  applicationInformer.Lister(),
+		applicationsSynced: applicationInformer.Informer().HasSynced,
+
+		releaseLister:  releaseInformer.Lister(),
+		releasesSynced: releaseInformer.Informer().HasSynced,
+
+		rolloutBlockLister: rolloutBlockInformer.Lister(),
+		rolloutBlockSynced: rolloutBlockInformer.Informer().HasSynced,
+
+		rolloutblockWorkqueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(),
+			"rolloutblock_controller_rolloutblocks",
+		),
+	}
+
+	glog.Info("Setting up event handlers")
+
+	releaseInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.enqueueReleaseBlock,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				controller.onUpdateRelease(oldObj, newObj)
+			},
+			DeleteFunc: controller.enqueueReleaseBlock,
+		})
+
+	applicationInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.enqueueApplicationBlock,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				controller.enqueueApplicationBlock(newObj)
+			},
+			DeleteFunc: controller.enqueueApplicationBlock,
+		})
+
+	rolloutBlockInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: controller.onAddRolloutBlock,
```

Comment:
I wonder if we shouldn't also listen to delete events here, and remove the rollout block being removed from all overrides in the apps and releases.

Currently, if a rollout block that was overriden gets deleted, and one tries to edit the app or release that overrode it, it'll error out with "rollout block with name foo/bar does not exist"